### DISCUSSION TITLE
How to make 3'rd party widgets using Native code?

### DISCUSSION THREAD

**Discussion by l-o-u-i-s-g** [beeware/toga#2927] [Q&A] [answered]
For simplicity as a fist implementation lets say  I want  to create a widget called  mybox for android and gtk by simply copying their respective box.py  to mybox.by
./core/src/toga/widgets/box.py
./gtk/src/toga_gtk/widgets/box.py
./android/src/toga_android/widgets/box.py

What  would I need to modify?

Class name from Box  to MyBox   in each files.
Implement MyBox instead of box in the core  but how would it know my source in the factory?
  self._impl = factory.MyBox(interface=self)

what else?
  
How would I structure / package that? 

Can I leverage the factory in place?

in the source I see  a factory.py  inside toga_gtk  and inside toga_android folders that point to the Box classes
which is also  in the __init__.py of the core. How could I make toga aware of my third party source?

I'm not trying to fork toga and add mybox.py to its source. Is there a clean way to package that as a third party?

**Comment by freakboy3742** [comment]
Right now, there isn't a built-in way to replicate what toga-core is doing around widget registration in a third-party widget.

The immediate workaround is to reproduce what toga-core is doing with factory handling in your own code - that is, define a factory function that checks `toga.platform.current_platform`, and based on that check, imports the appropriate backend implementation and returns an appropriate `_impl`.

However, longer term, this is something that I'd like to add as a baked-in capability of Toga. 

I can think I can think of two ways that this could be implemented.

1. Toga's core exposes a function that third-party modules can use to get a factory that can redirect to implementations as required.
2. We add [entry points](https://packaging.python.org/en/latest/specifications/entry-points/) for widgets, both for registering a widget, and for registering implementations of that widget. This way, installing a third-party widget package would make that widget available in the `toga.` namespace (or maybe in a `toga.ext.` extension namespace), and third-party widgets could invoke `factory()` to get the appropriate platform backend.

In the best case, Toga would "eat its own dogfood", and use the same mechanisms for built-in widgets that third party widgets use. 
We're already doing this with entry points to register backends; the interface that toga cocoa uses to register itself with toga core is accessible to third parties.

**Reply by freakboy3742** [reply, marked as answer]
I've opened #2928 to track the feature request.

**Reply by l-o-u-i-s-g** [reply]
Awesome! thanks!
